File: caper/samples/ruby/calc1_ast.rb
class Node
    def calc
    end
    def to_s
        ""
    end
end

class Expr < Node
end

class Term < Node
end

class Number < Node
    attr_accessor :number
    def initialize num
        @number = num
    end
    def calc
        @number
    end
    def to_s
        @number.to_s
    end
end

class AddExpr < Expr
    attr_accessor :lhs, :rhs
    def initialize x, y
        @lhs = x
        @rhs = y
    end
    def calc
        @lhs.calc + @rhs.calc
    end
    def to_s
        "(" + @lhs.to_s + " + " + @rhs.to_s + ")"
    end
end

class SubExpr < Expr
    attr_accessor :lhs, :rhs
    def initialize x, y
        @lhs = x
        @rhs = y
    end
    def calc
        @lhs.calc - @rhs.calc
    end
    def to_s
        "(" + @lhs.to_s + " - " + @rhs.to_s + ")"
    end
end

class TermExpr < Expr
    attr_accessor :term
    def initialize t
        @term = t
    end
    def calc
        @term.calc
    end
    def to_s
        @term.to_s
    end
end

class MulTerm < Term
    attr_accessor :lhs, :rhs
    def initialize x, y
        @lhs = x
        @rhs = y
    end
    def calc
        @lhs.calc * @rhs.calc
    end
    def to_s
        "(" + @lhs.to_s + " * " + @rhs.to_s + ")"
    end
end

class DivTerm < Term
    attr_accessor :lhs, :rhs
    def initialize x, y
        @lhs = x
        @rhs = y
    end
    def calc
        @lhs.calc / @rhs.calc
    end
    def to_s
        "(" + @lhs.to_s + " / " + @rhs.to_s + ")"
    end
end

class NumberTerm < Term
    attr_accessor :number
    def initialize num
        @number = num
    end
    def calc
        @number.calc
    end
    def to_s
        @number.to_s
    end
end

File: caper/samples/php/calc1_ast.php
<?php

class Node
{
    function calc()
    {
        return 0;
    }

    function __toString()
    {
        return "";
    }
}

class Expr extends Node
{
}

class Term extends Node
{
}

class Number extends Node
{
    public $number;

    function __construct($num)
    {
        $this->number = $num;
    }

    function calc()
    {
        return $this->number;
    }

    function __toString()
    {
        return (string)$this->number;
    }
}

class AddExpr extends Expr {
    public $lhs;
    public $rhs;

    function __construct($x, $y)
    {
        $this->lhs = $x;
        $this->rhs = $y;
    }

    function calc()
    {
        return $this->lhs->calc() + $this->rhs->calc();
    }

    function __toString()
    {
        return "(" . (string)$this->lhs . " + " . (string)$this->rhs . ")";
    }
};

class SubExpr extends Expr
{
    public $lhs;
    public $rhs;

    function __construct($x, $y)
    {
        $this->lhs = $x;
        $this->rhs = $y;
    }

    function calc()
    {
        return $this->lhs->calc() - $this->rhs->calc();
    }

    function __toString()
    {
        return "(" . (string)$this->lhs . " - " . (string)$this->rhs . ")";
    }
};

class TermExpr extends Expr
{
    public $term;

    function __construct($x)
    {
        $this->term = $x;
    }

    function calc()
    {
        return $this->term->calc();
    }

    function __toString()
    {
        return $this->term->__toString();
    }
};

class MulTerm extends Term
{
    public $lhs;
    public $rhs;

    function __construct($x, $y)
    {
        $this->lhs = $x;
        $this->rhs = $y;
    }

    function calc()
    {
        return $this->lhs->calc() * $this->rhs->calc();
    }

    function __toString()
    {
        return "(" . (string)$this->lhs . " * " . (string)$this->rhs . ")";
    }
};

class DivTerm extends Term
{
    public $lhs;
    public $rhs;

    function __construct($x, $y)
    {
        $this->lhs = $x;
        $this->rhs = $y;
    }

    function calc()
    {
        return $this->lhs->calc() / $this->rhs->calc();
    }

    function __toString()
    {
        return "(" . (string)$this->lhs . " / " . (string)$this->rhs . ")";
    }
};

class NumberTerm extends Term
{
    public $number;

    function __construct($x)
    {
        $this->number = $x;
    }

    function calc()
    {
        return $this->number;
    }

    function __toString()
    {
        return (string)$this->number;
    }
};
